crypto/src/base58/encodable.rs

```diff
@@ -20,17 +18,105 @@ pub trait Base58Encodable {
     {
         let mut buf = Vec::<u8>::new();
         to_writer(&mut buf, self).unwrap();
-        encode(buf).with_check_version(Self::VERSION_BYTE)
+        bs58::encode(buf).with_check_version(Self::VERSION_BYTE)
     }
 
-    fn from_base58<'a, I>(i: I) -> Result<Self, Error>
+    fn to_base58_string(&self) -> String
+    where
+        Self: Sized + Serialize,
+    {
+        self.to_base58().into_string()
+    }
+
+    fn from_base58<'a, I>(i: I) -> Result<Self, bin_prot::error::Error>
     where
         I: AsRef<[u8]>,
         Self: Sized + Deserialize<'a>,
     {
-        let bytes: Vec<u8> = decode(i).with_check(Some(Self::VERSION_BYTE)).into_vec()?;
+        let bytes: Vec<u8> = bs58::decode(i)
+            .with_check(Some(Self::VERSION_BYTE))
+            .into_vec()
+            .map_err(|e| bin_prot::error::Error::Custom {
+                message: format!("{:?}", e),
+            })?;
 
         // skip the first byte as this still contains the version byte
-        Ok(from_reader(&bytes[1..]).unwrap())
+        from_reader(&bytes[1..])
+    }
+}
+
+pub trait Base58EncodableHash {
+    /// This is the only part a custom implementation need provide.
+    /// Should be a constant from the base58_version_bytes.rs file corresponding
+    /// to the type.
+    const VERSION_BYTE: u8;
+    const MINA_VERSION_BYTE_COUNT: usize = 1;
+    const MINA_VERSION_BYTE: u8 = 1;
+    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>>
+    where
+        Self: AsRef<[u8; 32]>,
+    {
+        let mut buf = Vec::with_capacity(32 + Self::MINA_VERSION_BYTE_COUNT);
+        for _i in 0..Self::MINA_VERSION_BYTE_COUNT {
+            buf.push(Self::MINA_VERSION_BYTE);
+        }
+        for &b in self.as_ref() {
+            buf.push(b);
+        }
+        bs58::encode(buf).with_check_version(Self::VERSION_BYTE)
+    }
+
+    fn to_base58_string(&self) -> String
+    where
+        Self: AsRef<[u8; 32]>,
+    {
+        self.to_base58().into_string()
     }
+
+    fn from_base58<I>(i: I) -> Result<Self, bin_prot::error::Error>
+    where
+        I: AsRef<[u8]>,
+        Self: From<[u8; 32]>,
+    {
+        let bytes: Vec<u8> = bs58::decode(i)
+            .with_check(Some(Self::VERSION_BYTE))
+            .into_vec()
+            .map_err(|e| bin_prot::error::Error::Custom {
+                message: format!("{:?}", e),
+            })?;
+
+        // skip the bs58 version byte and mina bin_prot version bytes
+        let mut b32 = [0; 32];
+        b32.copy_from_slice(&bytes[(1 + Self::MINA_VERSION_BYTE_COUNT)..]);
+        Ok(b32.into())
+    }
+}
+
+#[macro_export]
+macro_rules! impl_hash_bs58_full {
```

Comment:
This is cool but I think a more idiomatic approach would be to make a custom derive macro

Reply:
actually having thought about this it is kind of nicer this way. @creativcoder do you have any thoughts on this?

Reply:
I believe it's preferable to do it this way unless the procedural macro is meant to part of public APIs, e.g.  https://github.com/rust-lang/portable-simd/blob/master/crates/core_simd/src/vector/float.rs#L208

Reply:
Yeah, I think if it simplifies the repetitive parts of code without sacrificing readability of the macro impl (which i think is simple enough here), we're fine with this.